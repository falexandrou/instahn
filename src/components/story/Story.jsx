import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { fetchStory } from 'actions/stories';

import './Story.scss';

/**
 * Renders a story item, showing its title, author & posted time
 */
@connect( (state, ownProps) => ({
  story: (state.stories.details[ownProps.storyId] || null)
}))
class Story extends React.Component {
  static propTypes = {
    storyId: PropTypes.number.isRequired,
    story: PropTypes.object,
    dispatch: PropTypes.func.isRequired,
  };

  componentDidMount() {
    const { dispatch, storyId } = this.props;
    dispatch( fetchStory(storyId) );
  }

  /**
   * @returns {String} the human-friendly date time representation
   */
  getStoryDateTime() {
    const { story: { time } } = this.props;

    if (! this.dateTime )
      this.dateTime = new Date( parseInt(time * 1000) );

    return `${this.dateTime.toLocaleDateString()} at ${this.dateTime.toLocaleTimeString()}`;
  }

  render() {
    const { story } = this.props;

    if (!story)
      return <div></div>;

    let storyDateTime = this.getStoryDateTime();

    return <article className="content story-item">
      <p>
        <a href={story.url} title={ story.title } target="_blank">
          { story.title }
        </a>
      </p>
      <div className="field is-grouped is-grouped-multiline">
        <div className="control">
          <div className="tags has-addons">
            <span className="tag">By</span>
            <span className="tag is-info">{story.by}</span>
          </div>
        </div>

        <div className="control">
          <span className="tag is-light">
            <time dateTime={ storyDateTime }>{ storyDateTime }</time>
          </span>
        </div>

        <div className="control">
          <div className="tags has-addons">
            <span className="tag">Score</span>
            <span className="tag is-primary">{ story.score }</span>
          </div>
        </div>
      </div>
    </article>;
  }
}

export default Story;