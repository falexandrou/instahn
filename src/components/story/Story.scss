.story-item {
  margin-bottom: 1.5em;
}