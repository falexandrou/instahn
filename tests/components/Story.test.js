import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Story from 'components/story/Story';
import { Provider } from 'react-redux';

import { createStore } from '../mocks/store';
import { storyMock } from '../mocks/fixtures';

const fakeStore = createStore({
  stories: {
    details: {
      [storyMock.id]: storyMock,
    },
  }
});

describe('Story Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Provider store={fakeStore}><Story storyId={storyMock.id} /></Provider>, div);
  });

  it('should match snapshot', () => {
    let component = <Provider store={fakeStore}><Story storyId={storyMock.id} /></Provider>;
    const output = renderer.create(component).toJSON();
    expect(output).toMatchSnapshot();
  });
});