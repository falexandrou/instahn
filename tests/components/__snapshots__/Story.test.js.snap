// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story Component should match snapshot 1`] = `
<article
  className="content story-item"
>
  <p>
    <a
      href="http://www.getdropbox.com/u/2/screencast.html"
      target="_blank"
      title="My YC app: Dropbox - Throw away your USB drive"
    >
      My YC app: Dropbox - Throw away your USB drive
    </a>
  </p>
  <div
    className="field is-grouped is-grouped-multiline"
  >
    <div
      className="control"
    >
      <div
        className="tags has-addons"
      >
        <span
          className="tag"
        >
          By
        </span>
        <span
          className="tag is-info"
        >
          dhouston
        </span>
      </div>
    </div>
    <div
      className="control"
    >
      <span
        className="tag is-light"
      >
        <time
          dateTime="2017-01-09 at 19:00"
        >
          2017-01-09 at 19:00
        </time>
      </span>
    </div>
    <div
      className="control"
    >
      <div
        className="tags has-addons"
      >
        <span
          className="tag"
        >
          Score
        </span>
        <span
          className="tag is-primary"
        >
          111
        </span>
      </div>
    </div>
  </div>
</article>
`;
